src/import/trace-event-json.ts

```diff
@@ -0,0 +1,229 @@
+ */
+function getActiveNodeArrays(profile: TraceEventJsonObject): Map<number, number[]> {
+  const map: Map<number, number[]> = new Map<number, number[]>()
+
+  // Given a nodeId, `getActiveNodes` gets all the parent nodes in reversed call order
+  const getActiveNodes = (id: number): number[] => {
+    if (map.has(id)) return map.get(id) || []
+
+    const node = profile.stackFrames[id]
+    if (!node) throw new Error(`No such node ${id}`)
+    if (node.parent) {
+      const array = getActiveNodes(node.parent).concat([id])
+      map.set(id, array)
+      return array
+    } else {
+      return [id]
+    }
+  }
+
+  Object.keys(profile.stackFrames).forEach(nodeId => {
+    const id = Number(nodeId)
+    map.set(id, getActiveNodes(id))
+  })
+
+  return map
+}
+
+/**
+ * Returns an array containing the time difference in microseconds between the previous
+ * sample and the current sample
+ */
+function getTimeDeltas(samples: Sample[]) {
+  const timeDeltas: number[] = []
+  let lastTimeStamp = Number(samples[0].ts)
+
+  samples.forEach((sample: Sample, idx: number) => {
+    if (idx === 0) {
+      timeDeltas.push(0)
+    } else {
+      const timeDiff = Number(sample.ts) - lastTimeStamp
+      lastTimeStamp = Number(sample.ts)
+      timeDeltas.push(timeDiff)
+    }
+  })
+
+  return timeDeltas
+}
+
+export function constructProfileFromJsonObject(
+  contents: TraceEventJsonObject,
+  samplesForPidTid: Sample[],
+  {profileBuilder}: ProfileBuilderInfo,
+) {
+  const activeNodeArraysById = getActiveNodeArrays(contents)
+
+  /**
+   * The json object format maintains an object of stack frames where the
+   * key is the frame id and the value is the stack frame object.
+   */
+  function getFrameById(frameId: string | number): StackFrame {
+    return contents.stackFrames[String(frameId)]
+  }
+
+  /**
+   * Wrapper function to get the active nodes for a given node id. We should
+   * always have active nodes for any given node.
+   */
+  function getActiveNodeIds(nodeId: number): number[] {
+    const activeNodeIds = activeNodeArraysById.get(nodeId)
+    if (!activeNodeIds) throw new Error(`No such node ID ${nodeId}`)
+    return activeNodeIds
+  }
+
+  // We need to leave frames in the same order that we start them, so we keep a stack
+  // of frames that are currently open
+  const frameStack: StackFrame[] = []
+
+  /**
+   * Enter a frame, pushing it to the top of the stack so that we keep track of what functions
+   * are currently being executed
+   */
+  function enterFrame(frame: StackFrame, timestamp: number) {
+    frameStack.push(frame)
+    profileBuilder.enterFrame(frameInfoForEvent(frame), timestamp)
+  }
+
+  /**
+   * Attempt to leave a frame. First we check if the frame matches what we expect to be the
+   * next thing to leave (top of the stack). If this is not the case we warn, and then leave
+   * the frame at the top of the stack
+   */
+  function tryToLeaveFrame(frame: StackFrame, timestamp: number) {
```

Comment:
There is some duplication here but I'm not sure if the juice is worth the squeeze...happy to refactor if you think otherwise